packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -180,12 +180,146 @@ class LineBreak extends TextBoundary {
   }
 }
 
+/// A text boundary that uses paragraphs as logical boundaries.
+///
+/// A paragraph is defined as the range between line terminators. If no
+/// line terminators exist then the paragraph boundary is the entire document.
+class ParagraphBoundary extends TextBoundary {
+  /// Creates a [ParagraphBoundary] with the text.
+  const ParagraphBoundary(this._text);
+
+  final String _text;
+
+  /// Returns the [TextPosition] representing the start position of the paragraph that
+  /// bounds the given `position`. The returned [TextPosition] is at the front of the leading
+  /// line terminator that encloses the desired paragraph.
+  @override
+  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
+    final Iterator<int> codeUnitIter = _text.codeUnits.iterator;
```

Comment:
Use indices to iterate since you'll need them anyways? I think that will make the implementation way simpler.